### DISCUSSION TITLE
Different JSON form representations for the same data

### DISCUSSION THREAD

**Discussion by agonzalezgaliana** [eclipse-emfcloud/emfcloud#157] [Theia Tree Editor]
Hi all,

We're working with the theia-tree-editor component. Our model-service class returns the different information to be used by JSON forms in order to show the details:

- UI Schema
- Node data
- JSON schema

During testing, we've found out that JSON forms shows the details with different layouting (even UI elements) depending on the information returned by the model-service.

The following is rendered when:
- UI Schema -> `undefined`
- Node data -> array of objects
- JSON schema -> the corresponding to the node

![Capture](https://user-images.githubusercontent.com/39406939/170648994-f0718d36-1333-4f22-bbe3-0bcfc8e3d5ae.PNG)
![Capture](https://user-images.githubusercontent.com/39406939/170649058-f702d1ef-2247-4d35-9c47-2891c7e34968.PNG)

However if the JSON schema returned is `undefined`, a table is rendered:

![Capture2](https://user-images.githubusercontent.com/39406939/170649234-e511a98c-e897-464d-b769-b147e0489349.PNG)
![Capture1](https://user-images.githubusercontent.com/39406939/170649249-c10b1722-97d9-4ae7-a6d7-db3f9262dd8e.PNG)

Can someone explain what's the reason for this behavior? I guess that JSON forms uses a different UI because the absence or presence of the schema.

Also because of this behavior, the add button triggers the `handleFormUpdate` method of the tree widget in one case (table case) and not in the other.

Any explanations to this?

Many thanks in advance.

**Comment by eneufeld** [comment]
Hi @agonzalezgaliana ,
JsonForms is using the provided JSONSchema to render a form. If a JSONSchema is not provided (undefined) it is derived from the provided data.
The first screen is rendered when the array contains 'complex' data, eg an object.
The second screen is rendered when the array only contains simple data, eg only strings.

Can you paste your schema and your data?
The button should behave the same in both cases and trigger a form update as in both cases a new entry should be added to the array.

Best,
Eugen

**Reply by agonzalezgaliana** [reply]
Hi @eneufeld thanks for the quick response.

The relevant schema snippet is the following:

```
               "import": {
			"$id": "#import",
			"title": "Import",
			"type": "object",
			"properties": {
				"eClass": {
					"const": "http://my_schema/1.0.0#//Import"
				},
				"document": {
					"type": "string"
				},
				"package": {
					"type": "string"
				},
				"prefix": {
					"type": "string"
				}
			},
			"additionalProperties": false,
			"required": ["document", "package", "prefix"]
		},
		"imports": {
			"$id": "#imports",
			"title": "Imports",
			"type": "object",
			"properties": {
				"eClass": {
					"const": "http://my_schema/1.0.0#//Imports"
				},
				"import": {
					"type": "array",
					"items": {
						"$ref": "#/definitions/import"
					}
				}
			},
			"additionalProperties": false,
			"required": ["import"]
		}
```
The data looks like the following:

```
{
	"import": [{
		"document": "Document1",
		"package": "Package1",
		"prefix": "Prefix"
	}]
}
```

The only thing it changes between the 2 scenarios is whether we return the JSON schema for the "Imports" object or `undefined`.

With regards the add button, in the table scenario the method is called, not in the other case.

**Reply by eneufeld** [reply]
Hi @agonzalezgaliana ,
I think the renderer tester doesn't handle the eClass `const` type correctly and thus takes the object list renderer instead of the table renderer. This would be a case for an issue in [jsonforms](https://github.com/eclipsesource/jsonforms/).
For the method not being called this needs to be debugged. 
Do you see any exceptions when pressing the button in the first scenario?

**Reply by agonzalezgaliana** [reply]
No errors or exceptions shown in the console :raised_eyebrow:

I'm debugging both scenarios. What I've seen till now is:

- Non working scenario: The `onClick` calls `addItem` with `path == 'import'` and `value == {}` (empty object as default value). Then I get to the `renderer.ts` class where the `addItem` method is being defined. Here I'm unable to continue since non of the following lines is being executed.
```
addItem: (path: string, value: any) => () => {  // The debugger gets to here but nothing inside the method is being called
    dispatch(
      update(path, array => {
        if (array === undefined || array === null) {
          return [value];
        }

        array.push(value);
        return array;
      })
    );
  }
```
- Working scenario: By clicking the button we jump directly to the `renderer.ts` class where the `addItem` method is being called (as opposite of the previous scenario).

**Reply by eneufeld** [reply]
hmm, that is strange.
So a breakpoint in line `if (array === undefined || array === null) {` is not called?
This looks like a jsonforms issue then. We will need to check.

**Reply by agonzalezgaliana** [reply]
> So a breakpoint in line `if (array === undefined || array === null) {` is not called? 

No, it's not called. I've retested it several times. Currently using JSON Forms "3.0.0-alpha.2"

**Reply by agonzalezgaliana** [reply]
I've opened [an issue in JSON Forms](https://github.com/eclipsesource/jsonforms/issues/1948)
